program/rust/src/c_oracle_header.rs

```diff
@@ -34,94 +34,137 @@ pub trait PythAccount: Pod {
+    pub num_:            u32,      // Current number of authorized publishers
+    pub num_qt_:         u32,      // Number of valid quotes for the last aggregation
+    pub last_slot_:      u64,      // Last slot with a succesful aggregation (status : TRADING)
+    pub valid_slot_:     u64,      // Second to last slot where aggregation was attempted
+    pub twap_:           PriceEma, // Ema for price
+    pub twac_:           PriceEma, // Ema for confidence
+    pub timestamp_:      i64,      // Last time aggregation was attempted
+    pub min_pub_:        u8,       // Minimum valid publisher quotes for a succesful aggregation
+    pub unused_1_:       i8,
+    pub unused_2_:       i16,
+    pub unused_3_:       i32,
+    pub prod_:           CPubkey, // Corresponding mapping account
+    pub next_:           CPubkey, // Next price account in the list
+    pub prev_slot_:      u64,     /* Second to last slot where aggregation was succesful
+                                   * (status : TRADING) */
+    pub prev_price_:     i64,       // Aggregate price at prev_slot_
+    pub prev_conf_:      u64,       // Confidence interval at prev_slot_
+    pub prev_timestamp_: i64,       // Timestamp of prev_slot_
+    pub agg_:            PriceInfo, // Last attempted aggregate results
+    pub comp_:           [PriceComponent; 32usize], // Publisher's price components
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct PriceComponent {
+    pub pub_:    CPubkey,
+    pub agg_:    PriceInfo,
+    pub latest_: PriceInfo,
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Pod, Zeroable)]
+pub struct PriceInfo {
+    pub price_:           i64,
+    pub conf_:            u64,
+    pub status_:          u32,
+    pub corp_act_status_: u32,
+    pub pub_slot_:        u64,
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Pod, Zeroable)]
+pub struct PriceEma {
+    pub val_:   i64,
+    pub numer_: i64,
+    pub denom_: i64,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Zeroable, Pod)]
+pub struct AccountHeader {
+    pub magic_: u32,
+    pub ver_:   u32,
+    pub type_:  u32,
+    pub size_:  u32,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Pod, Zeroable)]
+pub struct ProductAccount {
+    pub magic_:  u32,
+    pub ver_:    u32,
+    pub type_:   u32,
+    pub size_:   u32,
+    pub px_acc_: CPubkey,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct MappingAccount {
+    pub magic_:  u32,
+    pub ver_:    u32,
+    pub type_:   u32,
+    pub size_:   u32,
+    pub num_:    u32,
+    pub unused_: u32,
+    pub next_:   CPubkey,
+    pub prod_:   [CPubkey; 640usize],
+}
+
+// Unsafe impl because CPubkey is a union
+unsafe impl Pod for CPubkey {
+}
+unsafe impl Zeroable for CPubkey {
+}
+
+
+// Unsafe impl because product_list is of size 640 and there's no derived trait for this size
+unsafe impl Pod for MappingAccount {
+}
+unsafe impl Zeroable for MappingAccount {
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub union CPubkey {
```

Comment:
Both fields are currently used in the code. I feel like we can get to a safe derivation if we :
- Replace CPubkey by the solana sdk Pubkey 
- Update functions to only use `k1_` or `k8_`.
Let's do it in a future PR.

Reply:
Agree with the first suggestion, we should use the solana `Pubkey` type directly in a future PR. The second suggestion would just keep us using the ad-hoc unsafe functions we should scrap.